File: 0997 Find the Town Judge.py
class Solution:
    def findJudge(self, N: int, trust: List[List[int]]) -> int:
        trust_other = [0] * N
        be_trusted = [0] * N
        for pair in trust:
            trust_other[pair[0]-1] += 1
            be_trusted[pair[1]-1] += 1
        
        for i in range(N):
            if trust_other[i] == 0 and be_trusted[i] == N-1:
                return i+1
        return -1

''' Solution from lee215
    def findJudge(self, N, trust):
        count = [0] * (N + 1)
        for i, j in trust:
            count[i] -= 1
            count[j] += 1
        for i in range(1, N + 1):
            if count[i] == N - 1:
                return i
        return -1
'''

File: 0997 Find the Town Judge.java
class Solution {
    public int findJudge(int N, int[][] trust) {
        int[] trust_other = new int[N];
        int[] be_trusted = new int[N];
        for (int[] pair: trust) {
            ++trust_other[pair[0]-1];
            ++be_trusted[pair[1]-1];
        }
        for (int i = 0; i < N; ++i) {
            if (trust_other[i] == 0 && be_trusted[i] == N-1)
                return i+1;
        }
        
        return -1;
    }
}
/* Solution from lee215
Intuition:
Consider trust as a graph, all pairs are directed edge.
The point with in-degree - out-degree = N - 1 become the judge.

Explanation:
Count the degree, and check at the end.

Time Complexity:
Time O(T + N), space O(N)
*/
/*
    public int findJudge(int N, int[][] trust) {
        int[] count = new int[N+1];
        for (int[] t: trust) {
            count[t[0]]--;
            count[t[1]]++;
        }
        for (int i = 1; i <= N; ++i) {
            if (count[i] == N - 1) return i;
        }
        return -1;
    }
*/